File: server/Source/Grid.cs
namespace Server.Grid;

public class Offset(int x, int y)
{
    public int x = x;
    public int y = y;
}

public class Grid(int size, int offsetX, int offsetY)
{
    public int size = size;
    public Offset offset = new(offsetX, offsetY);

    public Grid Clone()
    {
        return new Grid(size, offset.x, offset.y);
    }
}

File: client/src/controllers/grid.ts
import type State from "../state";
import type Store from "../store";
import type GridView from "../views/grid";
import Controller from "./controller";

class GridController extends Controller<GridView> {
  constructor(store: Store, state: State, view: GridView) {
    super(store, state, view);

    this.view.listen("grid_change", ({ size, offsetX, offsetY }) => this.onchange(size, offsetX, offsetY));
    this.state.listen("grid_change", (grid) => this.view.set(grid.size, grid.offset.x, grid.offset.y));
  }

  private onchange(size: number, offsetX: number, offsetY: number) {
    const grid = { size, offset: { x: offsetX, y: offsetY } };

    this.state.setGrid(grid);
    this.store.send({
      type: "request_grid",
      grid,
    });
  }
}

export default GridController;
